### PR TITLE
Support multi-tenancy for incidents

### LINKED ISSUE #13426: A ResolveIncident request/command supports tenant validation
### Description

Support tenant validation when an incident is resolved.

:information_source: No changes on the Client side, as the tenant ownership is clear from the `IncidentRecord`, and the user's tenant id list is obtained from the gateway.

```[tasklist]
### Task breakdown
- [x] The `ResolveIncidentProcessor` resolves an incident only if the tenant that owns the incident is accessible by the user making the request.
- [x] The `ResolveIncidentProcessor` rejects a command if the tenant that owns the incident is **not** accessible by the user making the request.
- [x] ~Gateway provides a user's tenant id list with the broker request when a ResolveIncident RPC call is made~ (Done with #13989 and #14283)
```

### Blocked by
- #13345
- #13346

### PR DESCRIPTION
**Pull request by koevskinikola**
## Description

Ensure that Incidents are owned by a tenant, and only commands authorized for the owning tenant may resolve incidents.

:warning: Based on #14466 and should be merged after it.

## Related issues

closes #13426 
relates to #13346 

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [ ] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [ ] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

Other teams:
If the change impacts another team an issue has been created for this team, explaining what they need to do to support this change.
- [ ] [Operate](https://github.com/camunda/operate/issues)
- [ ] [Tasklist](https://github.com/camunda/tasklist/issues)
- [ ] [Web Modeler](https://github.com/camunda/web-modeler/issues)
- [ ] [Desktop Modeler](https://github.com/camunda/camunda-modeler/issues)
- [ ] [Optimize](https://github.com/camunda/camunda-optimize/issues)

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

### CODE REVIEW

**Comment by remcowesterhoud** [comment]
I've added a c[ommit that adds E2E tests ](https://github.com/camunda/zeebe/pull/14467/commits/1956e840376ebd145da6c1be7ba1c609079b56c0)

**Comment by koevskinikola** [comment]
@remcowesterhoud I finished all the tests here, so the PR is ready for review. Please have a look when you have the time.

File: protocol-impl/src/main/java/io/camunda/zeebe/protocol/impl/record/value/incident/IncidentRecord.java [hunk 1]
```diff
@@ -64,6 +68,7 @@ public IncidentRecord initFromProcessInstanceFailure(
     setBpmnProcessId(processInstanceEvent.getBpmnProcessIdBuffer());
     setProcessDefinitionKey(processInstanceEvent.getProcessDefinitionKey());
     setProcessInstanceKey(processInstanceEvent.getProcessInstanceKey());
+    setTenantId(processInstanceEvent.getTenantId());
```
Comment (remcowesterhoud) [inline protocol-impl/src/main/java/io/camunda/zeebe/protocol/impl/record/value/incident/IncidentRecord.java:71]: 🔧 Might as well remove this method as it's not used anywhere.

File: engine/src/main/java/io/camunda/zeebe/engine/processing/job/JobFailProcessor.java [hunk 2]
```diff
@@ -167,7 +167,8 @@ private void raiseIncident(final long key, final JobRecord value) {
         .setElementId(value.getElementIdBuffer())
         .setElementInstanceKey(value.getElementInstanceKey())
         .setJobKey(key)
-        .setVariableScopeKey(value.getElementInstanceKey());
+        .setVariableScopeKey(value.getElementInstanceKey())
+        .setTenantId(value.getTenantId());
```
Comment (remcowesterhoud) [inline engine/src/main/java/io/camunda/zeebe/engine/processing/job/JobFailProcessor.java:171]: ❌ You missed setting the tenant in one place: `JobThrowErrorProcessor.raiseIncident()`

File: engine/src/test/java/io/camunda/zeebe/engine/util/TestStreams.java [hunk 3]
```diff
@@ -13,6 +13,8 @@
 import static org.mockito.Mockito.mock;
 import static org.mockito.Mockito.when;
 
+import io.camunda.zeebe.auth.api.JwtAuthorizationBuilder;
+import io.camunda.zeebe.auth.impl.Authorization;
```
Comment (remcowesterhoud) [inline engine/src/test/java/io/camunda/zeebe/engine/util/TestStreams.java]: ❓ I'm a bit confused about [this](https://github.com/camunda/zeebe/pull/14467/commits/fb4c91bc7560783aa148132fedd1fa076824ca7b) commit. It mentions that you made the job stream tests tenant-aware, but all that's changes is a few imports were added.

File: engine/src/main/java/io/camunda/zeebe/engine/processing/job/JobFailProcessor.java (hunk 2 above)
Comment (remcowesterhoud) [reply engine/src/main/java/io/camunda/zeebe/engine/processing/job/JobFailProcessor.java:171]: This also means we don't have a test case for this. It'd be good to add this

**Review by remcowesterhoud — CHANGES_REQUESTED** [review]
Thanks @koevskinikola. I think you missed one scenario.

🤔 I also wonder if we need to add tests to verify that we can actually resolve the incidents for a non-default tenant. Currently we only check if we accept the command in the E2E tests. The behaviour has changed slightly so I think it'd be good to add a unit test for it.

File: engine/src/test/java/io/camunda/zeebe/engine/util/TestStreams.java (hunk 3 above)
Comment (koevskinikola) [reply engine/src/test/java/io/camunda/zeebe/engine/util/TestStreams.java]: I think this one got included with all the rebasing I did between branches. I'll clean it up.

File: engine/src/main/java/io/camunda/zeebe/engine/processing/job/JobFailProcessor.java (hunk 2 above)
Comment (koevskinikola) [reply engine/src/main/java/io/camunda/zeebe/engine/processing/job/JobFailProcessor.java:171]: :+1: Nice catch!

**Comment by koevskinikola** [comment]
@remcowesterhoud I implemented your review hints. Have a look when you have the time.

**Review by remcowesterhoud — APPROVED** [review]
LGTM

bors merge

**Pull request merged**